user/tests/wiring/power_management_long_running/power.cpp

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright (c) 2021 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "application.h"
+#include "unit-test/unit-test.h"
+
+#if !HAL_PLATFORM_POWER_MANAGEMENT
+#error "Unsupported platform"
+#endif // !HAL_PLATFORM_POWER_MANAGEMENT
+
+SYSTEM_MODE(SEMI_AUTOMATIC);
+UNIT_TEST_APP();
+
+#define SERIAL      Serial1
+
+namespace {
+
+constexpr uint32_t DISCHARGE_STATE1 = 0xc00f;
+constexpr uint32_t DISCHARGE_STATE2 = 0xd00f;
+constexpr uint32_t CHARGING_STATE_DISABLED_CHARGING = 0xbeef;
+constexpr uint32_t CHARGING_STATE_ENABLED_CHARGING = 0xdeed;
+constexpr uint32_t CHARGED_STATE = 0xc0fe;
+constexpr uint32_t DISCONNECTED_STATE = 0xd00d;
+constexpr uint32_t DISCONNECTED_STATE_SLEEP = 0xbead;
+constexpr uint32_t CONNECTED_STATE_SLEEP = 0xb007;
+uint32_t g_state;
+bool dischargeTested = false;
+
+void startup() {
+    SystemPowerConfiguration conf;
+    conf.feature(SystemPowerFeature::PMIC_DETECTION);
+    System.setPowerConfiguration(conf);
+}
+
+STARTUP(startup());
+
+void testDischarge() {
+    SCOPE_GUARD ({
+        dischargeTested = true;
+        if (Particle.connected()) {
+            SERIAL.println("> Disconnect from the cloud...");
+            Particle.disconnect();
+            SERIAL.println("> Turn off the Cellular modem.");
+            Cellular.off();
+        }
+        if (!Serial.isConnected()) {
+            SERIAL.println("\r\n> Please \"plug\" the USB and reconnect the Serial to continue the test.");
+            while (!Serial.isConnected());
+            delay(1s);
+            while (SERIAL.available() > 0) {
+                (void)SERIAL.read();
+            }
+        }
+    });
+    if (!dischargeTested) {
+        constexpr float DISCHARGE_TERMINATION_VCELL = 3.90f;
+
+        SERIAL.println("> Please \"unplug\" the USB. Press any key to continue when the hardware is set up.");
+        while (SERIAL.available() <= 0);
+        while (SERIAL.available() > 0) {
+            (void)SERIAL.read();
+        }
+
+        SERIAL.println("> Wait the power management settled...");
+        assertTrue(waitFor([]{
+            return System.batteryState() == BATTERY_STATE_DISCHARGING;
+        }, 10000));
+        assertTrue(System.powerSource() == POWER_SOURCE_BATTERY);
```

Comment:
All comparisons (except of course ones that do check a boolean) should use `assertEqual` otherwise it's impossible to tell what the actual value that caused the assertion to trigger was and we'll just see the result of the comparison - 0 (false).